rakelib/yard_builder.rb

```diff
@@ -0,0 +1,285 @@
+
+    load_releases[gem].first || "master"
+  end
+
+  def add_release gem, version
+    data = load_releases
+    data[gem] ||= []
+    data[gem] << version
+    store_releases data
+  end
+
+  def releases_file
+    gh_pages_dir + "_data" + "releases.yaml"
+  end
+
+  def load_releases
+    require "yaml"
+    YAML.load_file releases_file
+  end
+
+  def store_releases data
+    require "yaml"
+    sorted_data_pairs = data.each.sort.map do |gem, versions|
+      # sort versions in descending order
+      versions.uniq!.sort! do |a, b|
+        # Sort in descending order
+        Gem::Version.new(b.sub(/^v/,"")) <=> Gem::Version.new(a.sub(/^v/,""))
+      end
+      [gem, versions]
+    end
+    File.write releases_file, YAML.dump(Hash[sorted_data_pairs])
+  end
+
+  def current_git_commit path = "."
+    Dir.chdir path do
+      cmd "git rev-parse --short HEAD"
+    end
+  end
+
+  def determine_gems path = "."
+    Dir.chdir path do
+      raw_gems = cmd "git ls-files -- */*.gemspec"
+      raw_gems.split("\n").map { |gem| gem.split("/").first }.sort
+    end
+  end
+
+  def split_tag tag
+    fail "'tag' must be in the format <gem>/<version> Actual: #{tag}" unless tag.include?("/")
+    parts = tag.split("/")
+    fail "'tag' must be in the format <gem>/<version>. Actual: #{tag}" unless parts.length == 2
+    parts
+  end
+
+  def build_gem_docs gem, version, source_repo_dir, gh_pages_repo_dir
+    # remove any existing docs before we build new docs
+    safe_remove_dir(gh_pages_repo_dir + "docs" + gem + version)
+
+    # specify markup and provider in case a gem is missing this
+    markup = "--markup markdown --markup-provider redcarpet"
+    # specify the template path to get all customizations
+    template = "--template default --template-path #{gh_pages_repo_dir + "_yard_templates"}"
+    # # specify the .yardopts file
+    # yardopts = "--yardopts #{source_repo_dir + gem + ".yardopts"}"
+    # readme = "--readme #{source_repo_dir + gem + "README.md"}"
+    output_dir = "-o #{gh_pages_repo_dir + "docs" + gem + version}"
+    target_dir = "#{source_repo_dir + gem + "lib" + "**" + "*.rb"}"
+    # command_opts = [output_dir, markup, template, yardopts, readme, target_dir]
+
+    puts "cd #{source_repo_dir + gem}"
+    Dir.chdir(source_repo_dir + gem) do
+      cmds = [output_dir, markup, template]
+      cmd "yard --verbose #{cmds.join " "}"
+
+      # Manually move js and css assets.
+      # I'm sure there is an asset API to do this,
+      # but I can't figure it out and YARD makes no sense to me.
+      docs_dir = gh_pages_repo_dir + "docs"
+
+      FileUtils.remove_dir(docs_dir + "css") if Dir.exists?(docs_dir + "css")
+      FileUtils.remove_dir(docs_dir + "js") if Dir.exists?(docs_dir + "js")
+
+      FileUtils.mv(docs_dir + gem + version + "css", docs_dir)
+      FileUtils.mv(docs_dir + gem + version + "js", docs_dir)
+    end
+    fix_gem_docs gem, gh_pages_repo_dir
+  end
+
+  def fix_gem_docs gem, gh_pages_repo_dir
```

Comment:
If we go back and rebuild the old tags we will see this happen again. This can happen if we update the YARD template and we need all the old releases updated.

Reply:
Running `docs:rebuild_all` will need this.

Reply:
Running `docs:rebuild_tag[google-cloud-trace/v0.31.0]` will need this as well.